[ICDS Report] THR report backend new

Rohit25negi opened the pull request:
This is the new backend of the THR report.  The change is the split the services being delivered. 
Since all the THR indicators of THR report exists in SDR report. So this code changes the fetching layer of THR report to fetch the data from SDR report view instead of THR view.
Ticket: https://app.asana.com/0/1175081284767805/1175119480003914/f

FF: yes

I am yet to add another test case for the third report type. But this is good to initiate the review.
Best review, commit by commit

Rohit25negi committed:
Added the columns to the model

Rohit25negi committed:
udpated the aggregation

Rohit25negi committed:
added the view

Rohit25negi committed:
updated the fetching layer

Rohit25negi committed:
lynt fix

Rohit25negi committed:
fixed the errors

Rohit25negi committed:
added test case
s'

Rohit25negi committed:
added migration

czue reviewed (commented):
I didn't review the logic or tests line by line to ensure that all the values were correct but did try to wrap my head around the big picture of what's going on.

Holding off on approval to let you respond to comments. At a high level, the function that generates the excel seems quite complex and brittle, but probably addressing that well is out of scope for this PR?

Rohit25negi commented on custom/icds_reports/reports/take_home_ration.py:
```diff
@@ -97,3 +81,140 @@ def _format_infrastructure_data(data):
+                    'thr_25_days',
+                    'thr_distribution_image_count'
+                ]
+            columns += thr_columns
```
thanks.

Rohit25negi commented on custom/icds_reports/utils/__init__.py:
```diff
@@ -1101,22 +1101,35 @@ def create_thr_report_excel_file(excel_data, data_type, month, aggregation_level
+        data_start_row_diff = 2
+    else:
+        total_column_count = 11
+        data_start_row_diff = 1
```
hey yeah, I think we can find that with the length of the headers list. But I thought it will not be very clear why are these different. And this report type parameter is anyway being supplied from FE in THR report. and each report type "technically" corresponds to a different report generated with different columns. 
Thats why I though to use the report_type for this so that its clears which setting is for what

Rohit25negi commented on custom/icds_reports/utils/__init__.py:
```diff
@@ -1150,23 +1163,90 @@ def create_thr_report_excel_file(excel_data, data_type, month, aggregation_level
 
+        if report_type == 'beneficiary_wise':
+            if value in ('Total No. of Beneficiaries eligible for THR',
+                         'Total No. of beneficiaries received THR in given month'):
```
I wish there could be more easy way to do it But all these different settings are for different Excel format asked for different report type:
you can check the spec in this ticket: https://app.asana.com/0/1175081284767805/1157414140082868/f, 
These +1, +2 are hardcoded for these report formats as these formattings are very much belonged to each report type specifically. I though tried to make the sharable formatting between those reports without duplicating code

czue replied:
gotcha. might be nice to optionally add some context/comments so that someone looking at this later doesn't wonder the same thing... (although admittedly even after reading your comment I'm not clear exactly what they're doing... choosing a specific row in the sheet to put the data in?)

Rohit25negi committed:
Trimmed down the redundant code

Rohit25negi commented:
@czue  I have addressed the comments. This is good for second round of review. I am right now resolving the merge conflict. 
About the big Excel file generator function. This has been such that for all the reports we have. Each function is bonded to a separate formatted excel report that is why each function have some hardcoded numbers.  Maintaining which is becoming more tedious with time. I have thought some other approaches for this. Will soon sending out the proposal

Rohit25negi committed:
merged with master

Rohit25negi committed:
fixed the merged errors

Rohit25negi committed:
satisfy lynter

Rohit25negi committed:
merged with master

ShubhamBansal1997 commented on custom/icds_reports/migrations/0194_auto_20200528_1024.py:
```diff
@@ -0,0 +1,137 @@
+            field=models.IntegerField(null=True),
+        )
+    ]
+    operations.extend(get_view_migrations())
```
Since, we are adding new field, we need to remove this from the previous migration files??

czue reviewed (approved):
Thanks for the follow ups!

Glad to hear you have some ideas for improving the Excel functions - they're getting quite unwieldy!

ShubhamBansal1997 commented on custom/icds_reports/reports/take_home_ration.py:
```diff
@@ -97,3 +82,115 @@ def _format_infrastructure_data(data):
+
+    def get_beneficiary_wise_data(self, filters, order_by):
+
+        thr_days = '21' if self.month <= datetime(2020, 3, 1).date() else '25'
```
Please add a comment why are we taking 21 before that date and 25 after that date (2020-03-01) and set this as constant in the class constructor function and any specific reason of not directly using `date(2020,3,1)`

ShubhamBansal1997 commented on custom/icds_reports/reports/take_home_ration.py:
```diff
@@ -97,3 +82,115 @@ def _format_infrastructure_data(data):
+                   'Total No of Pictures taken by AWW']
+
+        if self.beta:
+            thr_column = 'thr_21_days' if self.month <= datetime(2020, 3, 1).date() else 'thr_25_days'
```
same as below

ShubhamBansal1997 commented on custom/icds_reports/reports/take_home_ration.py:
```diff
@@ -97,3 +82,115 @@ def _format_infrastructure_data(data):
         ]
+
+    def get_consolidated_data(self, filters, order_by):
+        headers = ['State', 'District', 'Block', 'Sector', 'Awc Name', 'AWW Name', 'AWW Phone No.',
```
`launched_column` is missing from headers

Rohit25negi replied:
That is intentional

ShubhamBansal1997 commented on custom/icds_reports/reports/take_home_ration.py:
```diff
@@ -97,3 +82,115 @@ def _format_infrastructure_data(data):
+        columns = ['state_name', 'district_name', 'block_name',
+                   'supervisor_name', 'awc_name',
+                   'aww_name', 'contact_phone_number',
+                   'num_launched_awcs', 'pw_thr_eligible',
```
`'num_launched_awcs'` is present in the columns but not in the headers

ShubhamBansal1997 replied:
@Rohit25negi this one also intentional?

ShubhamBansal1997 commented on custom/icds_reports/reports/take_home_ration.py:
```diff
@@ -1,21 +1,32 @@
-        def _format_infrastructure_data(data):
-            return data if data is not None else DATA_NOT_ENTERED
+        def _format_report_data(column, value, is_launched):
+            location_names = ['state_name', 'district_name', 'block_name', 'supervisor_name', 'awc_name']
```
nit: `location_headers` would be more meaningful var name

Rohit25negi committed:
fixed the migration

ShubhamBansal1997 reviewed (changes requested):
Left some points, some are important (other are leave it or take it)
- Please add tests cases for all the types
- Cross check excel files once for all the types

Rohit25negi committed:
moved the report types to const file

Rohit25negi committed:
used the context
fixed the test
cases

Rohit25negi committed:
satisfy lynter

Rohit25negi committed:
satify lynter

Rohit25negi committed:
fixed the test cases

Rohit25negi committed:
fixed test cases

Rohit25negi commented:
build passed: https://travis-ci.org/github/dimagi/commcare-hq/builds/694272485
